### PR TITLE
[Data] Add Logical Operator for `filter` and `flat_map`

### LINKED ISSUE #31936: [Datasets] Add logical operator for filter() and flat_map()
### Description

In this issue, we should implement the logical operator for `Dataset.filter()` and `Dataset.flat_map`.

### Use case

_No response_

### PR DESCRIPTION
**Pull request by amogkam**
Signed-off-by: amogkam <amogkamsetty@yahoo.com>

Adds logical operators for `filter()` and `flat_map()`

Closes https://github.com/ray-project/ray/issues/31936

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/logical/operators/map_operator.py [hunk 1]
```diff
@@ -26,8 +26,29 @@
...
+    Args:
+        name: Name for this operator. This is the name that will appear when inspecting
+            the logical plan of a Dataset.
+        input_op: The operator preceding this operator in the plan DAG. The outputs of
+            `input_op` will be the inputs to this operator.
+        block_fn: The transform function to apply to each input block to produce output
+            blocks.
+        target_block_size: The target size for blocks outputted by this operator.
+        fn: User provided UDF. `fn` is used as the first argument to `block_fn` if
+            provided.
```
Comment (c21) [inline python/ray/data/_internal/logical/operators/map_operator.py:39]: > `fn` is used as the first argument to `block_fn` if provided.

hmm not sure if I understand, but this is not what we are doing right? Can we say something like:  `fn` is called inside `block_fn`.

File: python/ray/data/dataset.py [hunk 2]
```diff
@@ -902,6 +904,17 @@ def transform(block: Block, fn: RowUDF[T, U]) -> Iterable[Block]:
...
+        if logical_plan is not None:
+            op = FlatMap(
+                input_op=logical_plan.dag,
+                block_fn=transform,
+                fn=fn,
+                compute=compute,
+                ray_remote_args=ray_remote_args,
+            )
+            logical_plan = LogicalPlan(op)
         return Dataset(plan, self._epoch, self._lazy)
```
Comment (c21) [inline python/ray/data/dataset.py:918]: return Dataset(plan, self._epoch, self._lazy, logical_plan)

**Review by c21 — COMMENTED** [review]
Thanks @amogkam for working on this!

File: python/ray/data/dataset.py (hunk 2 above)
Comment (c21) [reply python/ray/data/dataset.py:918]: To capture this bug, let's also add test for `flatmap()` and `filter()` in `test_execution_optimizer.py:test_e2e_optimizer_sanity()`. Thanks.

File: python/ray/data/_internal/logical/operators/map_operator.py (hunk 1 above)
Comment (amogkam) [reply python/ray/data/_internal/logical/operators/map_operator.py:39]: this is what's happening in `compute.py`

Comment (c21) [reply python/ray/data/_internal/logical/operators/map_operator.py:39]: oh it's actually second argument - https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/compute.py#L483 . But I don't think we need to give this level of detail on docstring, right?

File: python/ray/data/dataset.py (hunk 2 above)
Comment (amogkam) [reply python/ray/data/dataset.py]: good catch, updated and added test

File: python/ray/data/_internal/logical/operators/map_operator.py (hunk 1 above)
Comment (amogkam) [reply python/ray/data/_internal/logical/operators/map_operator.py]: sg, removed

**Review by c21 — APPROVED** [review]
LGTM

**Pull request merged**